python/hail/linalg/rowmatrix.py

```diff
@@ -0,0 +1,251 @@
+from hail.utils.java import Env, joption
+from hail.typecheck import *
+from hail.expr.expressions import expr_float64, matrix_table_source, check_entry_indexed
+
+row_matrix_type = lazy()
+
+
+class RowMatrix(object):
+    """Hail's row-distributed matrix of :py:data:`.tfloat64` elements.
+
+    .. include:: ../_templates/experimental.rst
+
+    Notes
+    -----
+    Row matrices must have fewer than :math:`2^31` columns.
+
+    Under the hood, row matrices are partitioned into group of rows.
+    """
+
+    def __init__(self, jrm):
+        self._jrm = jrm
+
+    @property
+    def n_rows(self):
+        """Number of rows.
+
+        Returns
+        -------
+        :obj:`int`
+        """
+        return self._jrm.nRows()
+
+    @property
+    def n_cols(self):
+        """Number of columns.
+
+        Returns
+        -------
+        :obj:`int`
+        """
+        return self._jrm.nCols()
+
+    @property
+    def shape(self):
+        """Shape of matrix.
+
+        Returns
+        -------
+        (:obj:`int`, :obj:`int`)
+           Number of rows and number of columns.
+        """
+        return self.n_rows, self.n_cols
+
+    @classmethod
+    @typecheck_method(entry_expr=expr_float64)
+    def from_entry_expr(cls, entry_expr):
+        """Creates a row matrix using a matrix table entry expression.
+
+        Parameters
+        ----------
+        entry_expr: :class:`.Float64Expression`
+            Entry expression for numeric matrix entries.
+
+        Returns
+        -------
+        :class:`RowMatrix`
+        """
+
+        check_entry_indexed('write_from_entry_expr/entry_expr', entry_expr)
+
+        mt = matrix_table_source('write_from_entry_expr/entry_expr', entry_expr)
+
+        #  FIXME: remove once select_entries on a field is free
+        if entry_expr in mt._fields_inverse:
+            field = mt._fields_inverse[entry_expr]
+            jrm = mt._jvds.toRowMatrix(field)
+        else:
+            field = Env.get_uid()
+            jrm = mt.select_entries(**{field: entry_expr})._jvds.toRowMatrix(field)
+
+        return cls(jrm)
+
+    @classmethod
+    @typecheck_method(path=str, partition_size=int)
+    def read_block_matrix(cls, path, partition_size):
```

Comment:
I like the name `from_block_matrix` better.

Reply:
I changed it to `read_from_block_matrix`. This is consistent with `BlockMatrix.from_entry_expr` and `BlockMatrix.write_from_entry_expr`. I want to reserve `RowMatrix.from_block_matrix` for a future command that takes a BlockMatrix rather than a path.

Reply:
Actually, I decided to add `RowMatrix.from_block_matrix` now as well.